TL-3.1-HSBM.py

```diff
@@ -0,0 +1,291 @@
+        self.plus_c = plus_c
+        self.min_split = min_split
+        self.selection_criteria = selection_criteria
+        self.X = X
+        self.adj = adj
+
+        # if cluster_method not in ["GMM", "KMeans", "Spherical-Kmeans"]:
+        #     msg = "clustering method must be one of [GMM, Kmeans, Spherical-Kmeans]"
+        #     raise ValueError(msg)
+
+        # if embed_method not in ["ASE", "LSE"]:
+        #     msg = "embedding method must be either ASE or LSE"
+        #     raise ValueError(msg)
+
+        labels = range(len(self.X))
+        labels = pd.DataFrame(labels)
+        labels.columns = ["inds"]
+        self.labels = labels
+
+        if root_inds is None:
+            print("No `root_inds` were input; assuming this is the root.")
+            root_inds = labels["inds"]
+        self.root_inds = root_inds
+
+    def fit(self):
+        while True:
+            all_k = []
+            for i, node in enumerate(self.get_lowest_level()):
+                node.fit_candidates()
+                node.select_model()
+                all_k.append(node.k_)
+            self.collect_labels()
+            if all_k.count(1) == len(all_k):
+                break
+
+        mask = self.labels.columns.str.endswith(('pred'))
+        self.labels = self.labels.loc[:, ~mask]
+
+    def fit_candidates(self):
+        if hasattr(self, "k_") and self.k_ == 0:
+            return
+
+        if self.is_root:
+            if self.ifembed is True:
+                X = self._embed()
+            else:
+                X = self.X
+        elif self.is_root is False:
+            if self.reembed is True:
+                n_components = self._choose_ncomponents()
+                X = self._embed(n_components=n_components)
+            else:
+                X = self.X
+
+        if np.linalg.norm(X) < 1:
+            X = X / np.linalg.norm(X)
+
+        # self.X_ = X
+        results = self._cluster(
+            X,
+            cluster_method="GMM",
+            n_init=self.n_init,
+        )
+
+        self.results_ = results
+
+    def _to_adjacency(self):
+        if self.adj is None:
+            pass
+        # TODO: convert data matrix e.g. distance or similarity matrix
+        # to adjacency matrix
+
+    def _choose_ncomponents(self):
+        max_dim = 0
+        for j, node in enumerate(self.get_lowest_level()):
+            dim = select_dimension(node.X, n_elbows=node.n_elbows)[0][-1]
+            if max_dim < dim:
+                max_dim = dim
+
+        return max_dim
+
+    def _embed(self, n_components=None):
+        if hasattr(self, "adj"):
+            adj = self.adj
+        else:
+            adj = self._to_adjacency()
+
+        embed_adj = pass_to_ranks(adj)
+        if self.plus_c:
+            embed_adj += 1 / adj.size
+
+        if n_components is None:
+            n_components = self.n_components
+
+        if self.embed_method == "ASE":
+            embedder = AdjacencySpectralEmbed(
```

Comment:
what do you think about having an `embed_kws` dictionary argument that can be passed into whichever embedding method? this would cut back on having so many params here